Add ability to suspend threads until a signal is called

reduz opened the pull request:
### Problem

For GDScript users that run code on threads, it often happens they want to `await` until something happens, the continue the thread execution.

Unfortunately, signals will most likely execute on the main thread, hence this means that the code being run on the thread will continue running on the main thread by default after the signal is awaiting for resumes.

To illustrate:

```GDScript
	# Code is running on a dedicated thread
	await button.pressed
	# Now code is running on the main thread
```

In the past, it has been discussed whether `await` should be smart about this and simply suspend the thread if running on one. The problem with this is, that users may also be willing to emit the signal that will resume from the thread itself and, at the time of awaiting, there is **no way for the interpreter to know on which thread the function will be resumed**.

Additionally, suspending the thread is a very different operation than awaiting. Awaiting saves the local function stack and returns immediately, while suspending stops the whole thread until another resumes it. Mixing and matching both seems, ultimately, undesired as they are not the same.

To solve this, a new static function is added to Thread by this pull request, which suspends a thread until a signal is emitted (no matter in which other thread).

### Solution

It works like this.

```GDScript
	# Suspends a thread until a button is pressed.
	Thread.suspend( button.pressed )
```

If code must do this and can run on both the main thread or a dedicated thread, it can do as follows:

```GDScript
	# Suspends a thread until a button is pressed.
	if (Thread.is_main_thread()):
		await button.pressed
	else:
		Thread.suspend( button.pressed )
```

For this, the `Thread.is_main_thread()` function (already existing in the internal Thread object) has been exposed to the engine API.

It also similarly works with returned values:

```GDScript
	# Suspends a thread until a button is pressed.
	if (Thread.is_main_thread()):
		val = await button.pressed
	else:
		val = Thread.suspend( button.pressed )
```

### Added warning

A warning was added if you try to await on a thread and resume in another, since in most cases users don't expect this to happen. It suggest using this new function instead. The warning can be disabled in the project settings if that behavior is _really_ intended.

### FAQ

**Q**: Are you sure this can't be done with await transparently?
**A**: No, please read again. There is no way for await to know beforehand how the function will be resumed. Only the user knows.

**Q**: Does it not make more sense to have an `await_thread` or something where the user will pass the argument?
**A**: I think its better to have a dedicated utility function for this. Not only the intention is more explicit, but additionally, as this makes it non exclusive to GDScript. It can be used from other languages that work with the engine.

**Q**: Thread safety wise, isn't connecting to a signal non-thread safe? Is this not a problem?
**A**: Yes, it is up to the user to protect and make sure nothing is connecting or emitting while the thread suspends. I think signals should be made thread safe, though at some point in the future.

**Q**: What happens if you await on a method ? How do you do it with suspend?
**A**: When GDScript awaits on a method, it actually awaits on the completed signal of the state object returned when awaiting. This class seems to not be exposed to the API, but can still be used, hence the following code should be valid:

```GDScript
func my_first_coroutine():
    await get_tree().process_frame
    return 42

func my_second_coroutine():
    var my_variable := Thread.suspend(my_first_coroutine().completed) as int
    print(my_variable) # Prints 42
```


<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

reduz committed:
Add ability to suspend threads until a signal is called

For GDScript users that run code on threads, it often happens they want to `await` until something happens,
the continue the thread execution.

Unfortunately, signals will most likely execute on the main thread, hence this means that the code being run on
the thread will continue running on the main thread by default.

It has been discussed whether await should be smart about this and simply suspend the
thread if running on one. The problem with this is, that users may also be willing to
emit the signal that will resume from the thread itself and, at the time of awaiting,
there is **no way for the interpreter to know on which thread the function will be resumed**.

Additionally, suspending the thread is a very different operation than awaiting. Awaiting saves the local
function stack and returns immediately, while suspending stops the whole thread until another resumes it.
Mixing and matching both seems, ultimately, undesired as they are not the same.

To solve this, a new utility function is added by this pull request, which
suspends a thread until a signal is emitted (no matter in which other thread).

It works like this.

```GDScript
	# Suspends a thread until a button is pressed.
	Thread.suspend( button.pressed )
```

If code must do this and can run on both the main thread or a dedicated thread,
it can do as follows:

```GDScript
	# Suspends a thread until a button is pressed.
	if (Thread.is_main_thread()):
		await button.pressed
	else:
		Thread.suspend( button.pressed )
```

For this, the `Thread.is_main_thread()` function (already existing in the internal Thread object)
has been exposed to the engine API.

**Q**: Are you sure this can't be done with await transparently?
**A**: No, please read again. There is no way for await to know beforehand how the function will be resumed. Only the user knows.

**Q**: Does it not make more sense to have an `await_thread` or something where the user will pass the argument?
**A**: I think its better to have a dedicated utility function for this. Not only the intention is more explicit, but additionally, as this makes it non exclusive to GDScript. It can be used from other languages that use the engine.

aaronfranke commented:
It makes sense to me, but why introduce a new word for it? Why not have a utility function called `thread_await` or `await_thread` instead of either a utility function called `thread_suspend` or a keyword?

reduz commented:
@aaronfranke to me awaiting on script is saving the call stack and then restoring it. Suspending a thread is stopping executing until something resumes it. Feels like different words for different concepts to me.

mihe commented:
Maybe it's better to have this be another static method on `Thread` instead, like `Thread.suspend_until`, to avoid adding more to the global namespace?

bruvzg reviewed (commented):
Is it usable across threads, `Node` signals will refuse to connect from a different thread group? Should there be an exception for this case?

But something like this seems to work fine (with `is_valid` added):
```gdscript
func thread_fun():
	var b = Button.new()
	b.text = "Test"
	call_deferred("add_child", b)
	thread_suspend(b.pressed)
	print("OK")
```

reduz commented:
@bruvzg makes sense, should probably add an optional argument for connecting on main thread and making it default.

sockeye-d commented:
Another function of await is being able to assign a variable the result of the signal. Would that be possible to implement with this method?

reduz commented:
@radiantgurl The problem of making it thread safe is the performance loss of having to lock a mutex every time you emit. 
This happens because inside of Object, signals are not allocated until connected (which saves a ton of memory). If you want to protect this properly, you have to lock both the signal and the signal map and use quite coarse locking, which is very undesirable and also quite prone to deadlocks. 

To do it properly, you most likely need to use raw pointers and a chunked list, together with fine grained atomics usage, so adding/removing elements is not as costly. Its doable, but its hard.

reduz commented:
@sockeye-d Yes, I fixed up the OP, you have to suspend with the completed signal, because when you await the function returns an actual state object. I think this should be better documented in the engine, but I added a note in the doc in the PR.

bruvzg commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,91 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+	}
+
+	virtual void call(const Variant **p_arguments, int p_argcount, Variant &r_return_value, Callable::CallError &r_call_error) const override {
+		semaphore->post();
```
Since signal can now be connected and called on the main thread, shouldn't it set the return value before `post`, otherwise `Thread::suspend` can return before it's set.

reduz replied:
That is a very good point, thanks for pointing this out! Will change it.

reduz replied:
In fact, it should be at the bottom so, if there is a call error, it wont post the semaphore.

MarianoGnu commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,93 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+			r_call_error.error = Callable::CallError::CALL_ERROR_TOO_MANY_ARGUMENTS;
+			r_call_error.argument = p_argcount;
+			r_call_error.expected = 1;
+			return;
```
Why returning an error instead of an Array with all arguments like `await` works with several arguments?

reduz replied:
Thats a good point, I feel it may be a little error prone to do this, but if this is the prefererd behavior it should be no problem to do it.

reduz replied:
I added it, simpler for consistency.

MarianoGnu commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,93 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+	Variant return_value;
+
+	Callable callable(memnew(CallableCustomSuspend(&semaphore, &return_value)));
+	if (p_connect_on_main_thread) {
```
shouldn't this be an automatic check using `if (!Thread::is_in_main_thread()) {` instead of  passing an argument and leave it to user's discression?

MarianoGnu replied:
Actually, if you put [this guard](https://github.com/godotengine/godot/pull/103896#discussion_r1989664550), this if condition is not even needed, you will always connect on main thread

reduz replied:
You would be surprised how oddly people uses this stuff and await, I think its better to do it explicitly and let the user choose to do this.

reduz commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,93 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+	}
+};
+
+Variant Thread::suspend(Signal p_resume_signal, bool p_connect_on_main_thread) {
```
I thought about it, I think you are right and will add this and eventually if anyone complains remove it.

MarianoGnu commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,17 @@
+				Suspend a thread until the signal provided in the [param resume_signal] argument is emitted. As connecting signals between threads is often discouraged (or blocked by thread guards), the [param connect_on_main_thread] is provided to perform the connection to the signal on the main thread.
+				Similarly to script 'await', if the signal provided is emitted with a single argument, then this function will return it. For multiple arguments, an array is returned.
+				As an additional note, if you intend to await on a script function instead of a signal, use the following syntax instead: [code] Thread.suspend( function().completed ) [/code].
+				This function will fail if called from the main thread.
```
we could add this codeblock for usage example (not sure about C# usage tho)
```
	[codeblock lang=gdscript]
	# Suspends a thread until a button is pressed.
	if (Thread.is_main_thread()):
		await button.pressed
	else:
		Thread.suspend( button.pressed )
	[/codeblock]
```

reduz replied:
Added

MarianoGnu commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,26 @@
+				As an additional note, if you intend to await on a script function instead of a signal, use the following syntax instead: [code] Thread.suspend( function().completed ) [/code].
+				This function will fail if called from the main thread.
+
+				Code example using this function with GDScript 'await'.
```
I dont think this header is needed, in html docs user has a picker to select what version of the code to look at (if available) and in editor i beliece gdscript version is always shown

![image](https://github.com/user-attachments/assets/3b07d59b-98c0-43e3-8174-f4b18e7a8a9e)

AThousandShips commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,98 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+	}
+
+public:
+	//for every type that inherits, these must always be the same for this type
```
```suggestion
```

AThousandShips commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,26 @@
+			<param index="1" name="connect_on_main_thread" type="bool" default="true" />
+			<description>
+				Suspend a thread until the signal provided in the [param resume_signal] argument is emitted. As connecting signals between threads is often discouraged (or blocked by thread guards), the [param connect_on_main_thread] is provided to perform the connection to the signal on the main thread.
+				Similarly to script 'await', if the signal provided is emitted with a single argument, then this function will return it. For multiple arguments, an array is returned.
```
```suggestion
				Similarly to script [code]await[/code], if the signal provided is emitted with a single argument, then this function will return it. For multiple arguments, an array is returned.
```

AThousandShips commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,26 @@
+				Similarly to script 'await', if the signal provided is emitted with a single argument, then this function will return it. For multiple arguments, an array is returned.
+				As an additional note, if you intend to await on a script function instead of a signal, use the following syntax instead: [code] Thread.suspend( function().completed ) [/code].
+				This function will fail if called from the main thread.
+
```
```suggestion
```

AThousandShips commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,26 @@
+				Code example using this function with GDScript 'await'.
+				[codeblock lang=gdscript]
+				# Suspends a thread until a button is pressed.
+				if (Thread.is_main_thread()):
```
```suggestion
				if Thread.is_main_thread():
```

AThousandShips commented on doc/classes/Thread.xml:
```diff
@@ -60,6 +66,26 @@
+				if (Thread.is_main_thread()):
+				    await button.pressed
+				else:
+				    Thread.suspend( button.pressed )
```
```suggestion
				    Thread.suspend(button.pressed)
```

dreed-sd commented:
> it is up to the user to protect and make sure nothing is connecting or emitting while the thread suspends.

If I have to acquire a lock to prevent other threads from emitting it in order to use `Thread.suspend`, then how would my thread ever resume, as I'd have the lock while suspended?

MarianoGnu commented:
There's also a chance that the object meant to emit the signal is destroyed, in that case the thread will remain suspended forever until the application ends?

reduz commented:
@dreed-sd If you are connecting to the main thread, you don't need to acquire a lock to prevent other threads from emitting. If you are connecting within threads (second argument being true), I can see this being a problem. 

For this, one possibility could be to pass an optional mutex as third argument to make the operation safe.

MarianoGnu commented:
> For this, one possibility could be to pass an optional mutex as third argument to make the operation safe.

Could each `Signal` instance have it's own mutex to be used only while connecting? or even having a satic/unique mutex
- With a mutex per signal, threads will only pause during conneciton of 2 threads are connecting to the same signal at the same time
- Wig a mutex per object, threads will only pause if during connection 2 threads are connecting to signals of the same owner object
- With a unique mutex, threads will only stop if they are connecting to any signal at the same time (which usually is unlikely because usually signals are only connected on the main thread)

radiantgurl commented on core/core_bind.cpp:
```diff
@@ -1405,6 +1409,98 @@ void Thread::set_thread_safety_checks_enabled(bool p_enabled) {
+		r_call_error.error = Callable::CallError::CALL_OK;
+
+		// Post after the return value has been set.
+		semaphore->post();
```
What happens if the callable is called two consecutive times? Will it take the second passed return array twice?

MarianoGnu replied:
The connecttion flag `CONNECT_ONE_SHOT` will prevent this from happening

radiantgurl commented on doc/classes/Thread.xml:
```diff
@@ -30,6 +30,12 @@
 				To check if a [Thread] is joinable, use [method is_started].
 			</description>
 		</method>
+		<method name="is_main_thread" qualifiers="static">
```
We should probably replace the main thread check in other areas with this, codestyle modification?

reduz commented:
Just for the record, I added thread safety to signals in #105453. I thought this would be very complex, but it turns out @AThousandShips did most of the work for this. Once that one is merged, it should be possible to merge this one too.

ValorZard commented:
To be clear, now that signals are thread safe, how does that change the implications of this PR?

reduz commented:
@ValorZard The thread unsafety of connecting to another signal in another thread (not the main thread) is now gone, so my [comment here](https://github.com/godotengine/godot/pull/103896#issuecomment-2717757890) can be disregarded.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -3211,6 +3211,9 @@
 			[/codeblock]
 		</member>
+		<member name="threading/warnings/warn_on_cross_thread_await" type="bool" setter="" getter="" default="true">
+			This setting forces a warning to emit when the script awaits for something in a thread, but the signal to resume it originated from another. This normally happens unintentionally but, if this behavior is truly intended, turn off this warning.
```
A script... awaits for "_something_"? But the signal to resume "_it_"...? From "_another_" what?

Also, these warnings are exposed as settings _because_ they may not always apply, so reminding that they can be disabled is quite redundant.
```suggestion
			If [code]true[/code], produces a warning when a script awaits for a signal in a thread, but the signal to resume the script originated from another thread. This behavior is usually unintentional.
```
That's with the assumption that it's correct. I haven't looked at prior discussion.
Does this warning only affect GDScript? If yes, it needs to be explicitly said, and the wording needs to be more specific. Also, in GDScript you can `await` for Callables so this may not be accurate. 

Also, the description really owes an explanation as to why this behaviour is undesirable. I assume it's because of confusion?

Mickeon reviewed (commented):
Mmrrgrmgrmm...
